libcontainer/specconv/spec_linux.go

```diff
@@ -306,6 +306,18 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
 			}
 		}
 	}
+
+	// Set the host UID that should own the container's cgroup.
+	// This must be performed after setupUserNamespace, so that
+	// config.HostRootUID() returns the correct result.
+	if v := spec.Annotations["org.opencontainers.runc.chown-cgroup"]; v == "true" {
```

Comment:
What if crun wants to add support for this? Can we use, say, `run.oci.cgroup.chown` or something like this?

@giuseppe PTAL

Reply:
@kolyshkin actually, `crun` already has this behaviour:

```
% jq '.linux.uidMappings, .linux.cgroupsPath, .process.user' < config.json
[
  {
    "containerID": 0,
    "hostID": 1000,
    "size": 1
  },
  {
    "containerID": 1,
    "hostID": 100000,
    "size": 65536
  }
]
"user.slice:runc:bar"
{
  "uid": 0,
  "gid": 0
}
% sudo crun --systemd-cgroup run test 
... # in separate shell
% ls -alin /sys/fs/cgroup/user.slice/test-test.scope/container |grep 1000
40139 drwxr-xr-x. 2 1000 1000 0 Sep 10 16:09 .
40141 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 cgroup.procs
40144 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 cgroup.subtree_control
40142 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 cgroup.threads
40176 -rw-r--r--. 1 1000 1000 0 Sep 10 16:09 memory.oom.group
```

Reply:
yes, crun always chowns the cgroup to root in the user namespace

Reply:
@kolyshkin with this in mind, what do you think about removing the annotation and just having this as the default behaviour in `runc` also?

Reply:
I'm definitely not a fan of using annotations for this; it feels like a band-aid / hack to work around missing functionality in the runtime spec. If there _would be_ an annotation, it should be one that's defined in the runtime spec (and image-spec?), which should describe how runtimes should process these annotations.

What are the consequences of setting this as default? Does it affect non-systemd-in-container situations?

Reply:
(apologies for going a bit orthogonal here)

I understand the "systemd in a container" use-case, but recall discussions from the past where some of those configuration changes needed were _beneficial_ for that use-case, but could have a _negative_ impact on other scenarios. I also recall other configuration changes (e.g. the `CONTAINER` env-var, which so far hasn't been described in the runtime spec) are needed to make systemd work properly inside a container.
Thinking out loud here, but would it be an idea of defining (e.g.) a concept of "profiles" in the runtime spec, that (when set) instructs the runtime to make the required changes to facilitate systemd? (`profile=systemd`)

Reply:
@thaJeztah as for non-systemd workloads, well, it means they can manage their own cgroups too.  I do not think it would break anything (never say never...)

As for changes to support systemd that break or impair other use cases, do you have examples?
In my case, this experimental version of runc, combined with the CRI-O user namespace support and worker nodes configured with cgroupsv2, is sufficient to run systemd-based container on OpenShift.

I am wrestling with some other problems with more advanced systemd features (e.g. *systemd-tmpfiles* not playing well with `emptyDir` mounts) but these can be worked around.

Reply:
> I'm definitely not a fan of using annotations for this; it feels like a band-aid / hack to work around missing functionality in the runtime spec. If there _would be_ an annotation, it should be one that's defined in the runtime spec (and image-spec?), which should describe how runtimes should process these annotations.

I will propose a clarification to the OCI runtime spec to specify that cgroups created by the OCI runtime should be owned by the container's uid 0.  At the very least some implementation notes/advice should be adopted into the OCI documents.

Reply:
runtime-spec PR: https://github.com/opencontainers/runtime-spec/pull/1123

Reply:
Thanks!

> As for changes to support systemd that break or impair other use cases, do you have examples?


Don't have direct examples, and I would have to dig into history for that.


> In my case, this experimental version of runc, combined with the CRI-O user namespace support and worker nodes configured with cgroupsv2, is sufficient to run systemd-based container on OpenShift.


@giuseppe may know more about the implementation, but it's possible that CRI-O and/or OpenShift automatically apply some configs and/or hooks to facilitate systemd inside a container.

There's a long history on systemd and containers with two camps (_"every container should have systemd as PID1"_ vs _"systemd doesn't belong in a container"_), which unfortunately led to a situation where either the user has to go through the (sometimes complicated) steps to make systemd work, or runtimes had to do preparations for this.

My personal opinion is that neither of those camps are "right", as it really depends on the use-case. In general, the default for _any_ container should be to have as little privileges as possible, and no _assumptions_ / _preparations_ should be made for a specific init system / tooling inside the container _by default_. _BUT_ I also think that it should be as easy as possible to be _able_ to use systemd (or other tools) in a container, and to be able to do so in a "portable" / "standard" way.

The OCI specifications are there to enhance portability; this is where I think "profiles" (or give it any other name) would be able to help. The OCI could define a "well known" profile for `systemd`, which would describe what preparations and/or changes to the container's spec runtimes are expected to make for containers using such profile (which could include permissions, environment variables, additional mounts and/or capabilities, etc.).

Reply:
I'd say this feature is needed more in general for enabling cgroup delegation when using a user namespace.

I don't think there should be any emphasis on systemd, since it is just an user of such feature.

For example on a cgroup v2 system:
```
$ podman --runtime /usr/bin/runc run --privileged --uidmap 0:1:1000 --rm fedora sh -c 'ls -lnd /sys/fs/cgroup; mkdir /sys/fs/cgroup/foo && echo created'
drwxr-xr-x. 2 65534 65534 0 Sep 13 08:02 /sys/fs/cgroup
mkdir: cannot create directory '/sys/fs/cgroup/foo': Permission denied
```

crun automatically chowns the cgroup to root in the userns:
```
$ podman --runtime /usr/bin/crun run --privileged --uidmap 0:1:1000 --rm fedora sh -c 'ls -lnd /sys/fs/cgroup; mkdir /sys/fs/cgroup/foo && echo created'
drwxr-xr-x. 2 0 0 0 Sep 13 08:02 /sys/fs/cgroup
created
```

Reply:
@giuseppe indeed, systemd workload is just one notable use case.

Separately, your input over at https://github.com/opencontainers/runtime-spec/pull/1123 would be appreciated.

Reply:
If this should be set unconditionally (and if it can be done without risk of introducing breaking changes), I'm definitely +1 on doing so; I agree that  https://github.com/opencontainers/runtime-spec/pull/1123 would be good for that (to define that in the spec).

Reply:
If we are going to enable chowning by default, we will need to have a way to opt-out for that.

I guess skipping chowning when the mount is ro (https://github.com/opencontainers/runtime-spec/pull/1123#discussion_r707393997 , suggested by @giuseppe) would be fine.

Reply:
I have updated the PR to remove the annotation, and only chown the cgroup when the cgroupfs is mounted `rw`.